src/Contrib.KafkaFlow.Outbox.SqlServer/SqlServerOutboxBackend.cs

```diff
@@ -0,0 +1,95 @@
+﻿using Confluent.Kafka;
+using Dapper;
+using KafkaFlow.SqlServer;
+using Microsoft.Extensions.Options;
+using System.Data.SqlClient;
+using System.Text.Json;
+
+namespace KafkaFlow.Outbox.SqlServer;
+
+public class SqlServerOutboxBackend : IOutboxBackend
+{
+    private readonly SqlServerBackendOptions _options;
+
+    public SqlServerOutboxBackend(IOptions<SqlServerBackendOptions> options)
+        => _options = options?.Value ?? throw new ArgumentNullException(nameof(options));
+
+    public async ValueTask Store(TopicPartition topicPartition, Message<byte[], byte[]> message, CancellationToken token = default)
+    {
+        var sql = """
+            INSERT INTO [outbox] ([topic_name], [partition], [message_key], [message_headers], [message_body])
+            VALUES (@topic_name, @partition, @message_key, @message_headers, @message_body);
+            """;
+
+        using var conn = new SqlConnection(_options.ConnectionString);
+
+        var rawHeaders =
+            message.Headers == null
+            ? null
+            : JsonSerializer.Serialize(message.Headers.ToDictionary(x => x.Key, x => x.GetValueBytes()));
+
+        var res = await conn.ExecuteAsync(sql, new
+        {
+            topic_name = topicPartition.Topic,
+            partition = topicPartition.Partition.IsSpecial ? null : (int?)topicPartition.Partition.Value,
+            message_key = message.Key,
+            message_headers = rawHeaders,
+            message_body = message.Value
+        }).ConfigureAwait(false);
+
+    }
+
+    public async ValueTask<OutboxRecord[]> Read(int batchSize, CancellationToken token = default)
+    {
+        var sql = """
+            DELETE FROM [outbox]
+            OUTPUT DELETED.*
+            WHERE
+                [sequence_id] IN (
+                    SELECT TOP (@batch_size) [sequence_id] FROM [outbox]
+                    ORDER BY [sequence_id]
+                );
```

Comment:
@AlexeyRaga I have a question/consideration to think about here...

This will output the records from the outbox in the sequence_id order. This _isnt necessary_ the same order as they went into the outbox transactionally. Its likely... but not guaranteed. If we want it to be transactionally ordered we would need to use the `rowversion` data type. However; @RobThree raised that the `rowversion` column (not the type) is currently an `INT` whereas the `rowversion` type in SQL Server is 8 bytes (kind of like an unsigned big int... but not :smile:).

Do we need the `rowversion` column to be a change counter on that row for the data? Or can we replace it?

If we need to keep it, I'd recommend adding a secondary column on the `outbox` table with a `rowversion` type and ordering by that. We wouldnt need to select it and could retain the `rowversion` column that is currently doing the change count.

Example of out of order sequence_ids:

| Thread A| Thread B |
|--------------|--------------|
|BEGIN TRAN| BEGIN TRAN|
|INSERT INTO outbox||
||INSERT INTO outbox|
||COMMIT TRAN|
|COMMIT TRAN|

Thread A got sequence_id 1
Thread B got sequence_id 2
Thread B was committed to the database first

Reply:
@kchenery speaking about ordering and the edge cases, I think you are correct, BUT let's think about whether it is of any importance for us...

Here is a train of thoughts...

####  ❓ _When_ this can happen?
It can only happen when two _separate_ and _parallel_ actors insert into the outbox and their transactions overlap in time.

#### ❓  Is it important? Do we care?
Since these are two independent activities, their ordering is co-incidental anyway. Assume that there was no outbox and was just pure Kafka. Do we have any guarantee about messages that are being independently sent from different actors? 

We probably don't. Network, internal messages buffering in `librdkafka`, brokers themselves... And even logically, there is no order between independent parallel producing messages.

This means that _in reality_ the messages order is established _at the receiver side_ where the receiver is either Kafka itself, or our outbox proxy.

We also remove messages from the outbox and never can replay from it for determinism, so even if we get into a situation that messages from two separate actors that came at almost the same time got delivered "out of order", this gets on-observable for the rest of the world (due to its coincidental nature) and can't have any consequences?

If my reasoning above is correct then we are probably OK with the current implementation?